src/thing_creature.c

```diff
@@ -980,337 +1116,404 @@ void first_apply_spell_effect_to_thing(struct Thing *thing, SpellKind spell_idx,
+void reapply_spell_effect_to_thing(struct Thing *thing, SpellKind spell_idx, long spell_lev, int slot_idx)
 {
-    struct CreatureControl* cctrl = creature_control_get_from_thing(thing);
+    struct CreatureControl *cctrl = creature_control_get_from_thing(thing);
     if (spell_lev > SPELL_MAX_LEVEL)
+    {
         spell_lev = SPELL_MAX_LEVEL;
-    struct CastedSpellData* cspell = &cctrl->casted_spells[idx];
+    }
     // This pointer may be invalid if spell_idx is incorrect. But we're using it only when correct.
-    struct SpellConfig* spconf = get_spell_config(spell_idx);
-    const struct MagicStats* pwrdynst = get_power_dynamic_stats(spconf->linked_power);
+    struct SpellConfig *spconf = get_spell_config(spell_idx);
+    const struct MagicStats *pwrdynst = get_power_dynamic_stats(spconf->linked_power);
+    GameTurnDelta duration;
+    // If not linked to a keeper power, use the duration set on the spell, otherwise use the strength or duration of the linked power.
+    if (spconf->linked_power == PwrK_None)
+    {
+        duration = spconf->duration;
+    }
+    else if (pwrdynst->duration == 0)
+    {
+        duration = pwrdynst->strength[spell_lev];
+    }
+    else
+    {
+        duration = pwrdynst->duration;
+    }
+    // Only reset spell duration if the creature can still be affected by at least one spell flags set.
+    if (set_thing_spell_flags(thing, spell_idx, duration, spell_lev))
+    {
+        struct CastedSpellData *cspell = &cctrl->casted_spells[slot_idx];
+        cspell->duration = duration;
+        if (spconf->aura_effect != 0)
+        {
+            cctrl->spell_aura = spconf->aura_effect;
+            cctrl->spell_aura_duration = spconf->duration;
+        }
+    }
+    return;
+}
 
-    short duration;
-    if (spconf->linked_power == 0)
+void apply_spell_effect_to_thing(struct Thing *thing, SpellKind spell_idx, long spell_lev)
+{
+    struct CreatureControl *cctrl = creature_control_get_from_thing(thing);
+    if (creature_control_invalid(cctrl))
+    {
+        ERRORLOG("Invalid creature %s index %d tried to accept spell %s", thing_model_name(thing), (int)thing->index, spell_code_name(spell_idx));
+        return; // Exit the function, control is invalid.
+    }
+    struct SpellConfig *spconf = get_spell_config(spell_idx);
+    if (spell_config_is_invalid(spconf))
     {
-        duration = spconf->duration;
+        ERRORLOG("Spell %s config is invalid", spell_code_name(spell_idx));
+        return; // Exit the function, spell config is invalid.
     }
-    else
-    if (pwrdynst->duration == 0)
+    if (creature_is_immune_to_spell_flags(thing, spconf->spell_flags))
     {
-        duration = pwrdynst->strength[spell_lev];
+        SYNCDBG(7, "Creature %s index %d is immune to all spell flags %d set on %s", thing_model_name(thing), (int)thing->index, (uint)spconf->spell_flags, spell_code_name(spell_idx));
+        return; // Exit the function, creature is immune to all spell flags set on spell_idx.
```

Comment:
This seems wrong. It drills down to `creature_is_immune_to_spell_flags_f`:
```c
/* Returns if the creature kind is immune to a spell flags.
 * @param thing The thing to be checked.
 * @param spell_flags The spell flags to be checked. */
TbBool creature_is_immune_to_spell_flags_f(const struct Thing *thing, unsigned long spell_flags, const char *func_name)
```
`a spell flags.` is incorrect grammar. 
And also, you are pretending this is a single flag. There is the `all_flags_are_set` function. And if you do want that, name the function 'creature_is_immune_to_all_spell_effects' or something.

Reply:
Ok, you tested this to work too? And it will continue to apply just the spells it can be effected by?